dir.go

```diff
@@ -405,7 +405,8 @@ func (d *Directory) sync() error {
 		}
 	}
 
-	// TODO: Should we clean the cache here?
+	// Clear entries cache
+	d.entriesCache = make(map[string]FSNode)
```

Comment:
What happens if there is an error in the loop above and we already process half of this? (Maybe an error already signals corruption and the cache state is no longer important; I'm not sure what are the expectations here.)

Reply:
If there is an error in the loop, then the function will return and not clear the cache entries.  If possible these can be retried later.  Only if the loop completes successfully are the cache entries cleared, which is safe because all the entries have been updated.  This avoids unnecessary updates in the future due to nodes that are still in the cache but are already up-to-date.

Reply:
I get the rationale of why do we clear the cache in general, my question was more specific to the case where, say, I have 2 entries in the cache, `A` and `B`, then the loop processes correctly `A` and fails at `B`, but the cache still has both `A` and `B` in it because we exited early inside the loop (when `updateChild` returned an error) and we didn't reach this line. The crux of my question then is: a future call to `sync` will update again `A` which we didn't want, potentially overwriting a newer entry with same name, is this an issue? If so, do we want to fix it, add a `FIXME`, comment it as part of the API?

Reply:
As far as I can tell, a future call to sync in the case described above is harmless, and will only cause an unnecessary update of the same data that is already present.  It will not overwrite a newer entry with the same name, as the creation/update of the newer entry will also update the cache.

However, in the case of an unnecessary update following a sync failure, the timestamp of the entry may get updated.  So, I now delete each entry from the cache as soon as it is synced.